### PR TITLE
Add revert command for the initialize step

### PR DESCRIPTION
**Pull request by marksliva**
The revert command is only being tested after running a full initialize,
however it should be idempotent enough to handle reverting from any substep or
failure encountered during initialize. The revert process includes:
- delete segment and standby datadirs
- stop agents
- delete master datadir
- stop hub
- delete gpupgrade state directory

### CODE REVIEW

File: test/revert.bats [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        --source-bindir="$GPHOME/bin" \
+        --target-bindir="$GPHOME/bin" \
+        --source-master-port="${PGPORT}" \
+        --temp-port-range 6020-6040 \
+        --disk-free-ratio 0 \
+        --verbose 3>&-
+
+    # parse config.json for the datadirs
+    local target_datadirs=$(jq ".Target.Primaries[].DataDir" "${GPUPGRADE_HOME}/config.json")
+
```
Comment (kalensk) [inline test/revert.bats:56]: Should this be exposed via `gpupgrade config show --target-datadirs` instead? That way it follows the idiom we have been using throughout the BATs tests.

Comment (marksliva) [reply test/revert.bats:56]: Potentially. However, we would have to add a new flag to the config that returns all target datadirs, since the only datadir flag `target-datadir` returns only the masterDataDir
```
	case "target-datadir":
		resp.Value = s.Target.MasterDataDir()
```
Other tests that have needed the target datadirs have queried the database for the datadirs. However, the reason we did not do that was because the target cluster is shut down at the end of initialize.

Comment (kalensk) [reply test/revert.bats:56]: Wouldn't using a different variable be easier than a flag. That is `target-datadir` and `target-directories` or something?

Comment (marksliva) [reply test/revert.bats:56]: I'm not sure what you mean by easier. Either way, we'd have to call config with a specific flag or arg regardless.

File: hub/delete_master_data_directory.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	_, err := utils.System.Stat(masterDataDir)
+	if err != nil {
+		if xerrors.Is(err, os.ErrNotExist) {
+			return nil
+		}
+		mErr = multierror.Append(mErr, err)
+		return mErr
+	}
+
+	for _, fileName := range agent.PostgresFiles {
```
Comment (berlin-ab) [inline hub/delete_master_data_directory.go:25]: This seems to be duplication between the agent code. Also, the `hub` package shouldn't have a dependency on the `agent` for this work.

File: test/revert.bats (hunk 1 above)
Comment (marksliva) [reply test/revert.bats:56]: We checked with product, and they don't think any customers need this, so there's not much reason to add it to the binary. In our test, it is a one-liner, so we'll stick with that for now.

Comment (kalensk) [reply test/revert.bats:56]: Sure, but we are deviating from previous patterns and I guess I have a strong preference for consistency.

For example, a previous [PR](https://github.com/greenplum-db/gpupgrade/pull/179) introduced a new [gpupgrade config](https://github.com/greenplum-db/gpupgrade/commit/1fca144a08b470c9bcfb83f5122e5a619ea19cdd#diff-ab4ba84babe11ee4bb488285f23e46f8
) option simply for testing such as in [pg_ugprade.bats]( https://github.com/greenplum-db/gpupgrade/blob/140885913865233367ce55cf0b36b1288ad4ccd2/test/pg_upgrade.bats#L59) and other places. 

I guess my comment is pointing out that now we have two different ways of getting information used in our tests: 1) gpupgrade config show, and 2) using jq to parse the config file.

File: agent/delete_data_directories.go [hunk 3]
```diff
@@ -7,6 +7,10 @@ import (
 	"github.com/greenplum-db/gpupgrade/idl"
 	"github.com/greenplum-db/gpupgrade/upgrade"
 )
+// todo: needs to go in its own file
+func (s *Server) DeleteDirectory(ctx context.Context, in *idl.DeleteDirectoryRequest) (*idl.DeleteDirectoryReply, error) {
+	return nil, nil
+}
```
Comment (kalensk) [inline agent/delete_data_directories.go:13]: I know this is still in progress so pardon this comment if its already being worked on. Similar to delete_data_directories.go we will probably want to add an additional safety check to ensure we are deleting the appropriate directory.

File: idl/hub_to_agent.proto [hunk 4]
```diff
@@ -36,8 +37,12 @@ message UpgradePrimariesReply {}
 message DeleteDirectoriesRequest {
   repeated string datadirs = 1;
 }
+message DeleteDirectoryRequest {
+  string directory = 1;
+}
```
Comment (kalensk) [inline idl/hub_to_agent.proto:42]: I think we should consider calling this `DeleteStateDirectory` to emphasize that it won't/shouldn't delete any arbitrary directory for safety/security reasons. Rather it should check and make sure to delete only sate directories.

**Comment by dkrieger-pvtl** [comment]
> The revert command is only being tested after running a full initialize,
> however it should be idempotent enough to handle reverting from any substep or
> failure encountered during initialize.

I think you mean to say that this story is only handling revert after a full initialize has completed, right?  Recovering from a failure mid-initialize is more work(might have to stop postgres processes, etc), and will be handled later.  As will saving off the state directory.

Note the gpAdminLogs are not touched by this PR, but I suppose we consider those system logs and hence we do not need to delete them.

File: agent/delete_directories_test.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	"reflect"
+	"testing"
+
+	"github.com/greenplum-db/gp-common-go-libs/testhelper"
+	"github.com/greenplum-db/gpupgrade/agent"
+	"github.com/greenplum-db/gpupgrade/idl"
+	"github.com/greenplum-db/gpupgrade/utils"
+)
+
+func TestDeleteDirectories(t *testing.T) {
```
Comment (dkrieger-pvtl) [inline agent/delete_directories_test.go:15]: Consider creating actual directories here to test...you can use a temp dir and create the files you need underneath.  A strict mock-style test pattern would mock out the system calls, but in this case, you can test the actual functionality directly in a straightforward manner.

File: test/revert.bats (hunk 1 above)
Comment (jchampio) [reply test/revert.bats:56]: > We checked with product, and they don't think any customers need this, so there's not much reason to add it to the binary.

`config show` isn't for users, it's for us (and potentially support). It exists to help us partially decouple hub internals from the tests that rely on configuration values like this.

File: upgrade/directories.go [hunk 6]
```diff
@@ -56,3 +59,30 @@ func TempDataDir(datadir, segPrefix string, id ID) string {
 
 	return filepath.Join(dir, newBase)
 }
+
+func DeleteDirectories(directoriesToDelete []string, filesThatMustExist []string) *multierror.Error {
+	var mErr *multierror.Error
+	for _, segDataDir := range directoriesToDelete {
+
+		var postgresFilesError *multierror.Error
```
Comment (kalensk) [inline upgrade/directories.go:67]: I might be misreading, but why the need for this variable rather than keep things simple and just use `mErr`? Why separate the errors to just combine them in `mErr` anyway?

File: upgrade/directories_test.go [hunk 7]
```diff
@@ -48,3 +55,112 @@ func ExampleTempDataDir() {
...
+				return expected
+			}
+			return nil
+		}
+
+		statCalls := 0
+		utils.System.Stat = func(name string) (os.FileInfo, error) {
+			statCalls++
+			return nil, nil
+		}
```
Comment (kalensk) [inline upgrade/directories_test.go:139]: I think this test should be stronger. Something like verifying that stat is being called for on the inputs which in this case is foo and bar for each of the number of directories.

File: hub/revert.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+			gplog.Error(fmt.Sprintf("revert: %s", err))
+		}
+	}()
+
+	err = DeleteSegmentAndStandbyDirectories(s.agentConns, s.Config.Target)
+	if err != nil {
+		return err
+	}
+
+	// todo: DeleteStateDirectories()
```
Comment (dkrieger-pvtl) [inline hub/revert.go:35]: Did you mean to fix this as part of this PR?  

One potential issue if it is enabled: how do we know no call after this one will, in fact, attempt to read/write the state directory?

File: agent/delete_directories.go [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+	mErr := upgrade.DeleteDirectories([]string{in.Directory}, stateDirectoryFiles)
+	return &idl.DeleteStateDirectoryReply{}, mErr.ErrorOrNil()
+}
+
+func (s *Server) DeleteDirectories(ctx context.Context, in *idl.DeleteDirectoriesRequest) (*idl.DeleteDirectoriesReply, error) {
+	gplog.Info("got a request to delete data directories from the hub")
+
+	mErr := upgrade.DeleteDirectories(in.Datadirs, postgresFiles)
+	return &idl.DeleteDirectoriesReply{}, mErr.ErrorOrNil()
```
Comment (kalensk) [inline agent/delete_directories.go:26]: Can these be renamed to `DeleteDataDirectories` for clarity? I find `DeleteDirectories` to be ambiguous when trying to follow the code.

File: hub/revert.go (hunk 8 above)
Comment (kalensk) [reply hub/revert.go:35]: > One potential issue if it is enabled: how do we know no call after this one will, in fact, attempt to read/write the state directory?

Aren't we in control of that? That is, wouldn't that be a developer bug?

File: idl/hub_to_agent.proto [hunk 10]
```diff
@@ -36,8 +37,12 @@ message UpgradePrimariesReply {}
 message DeleteDirectoriesRequest {
```
Comment (dkrieger-pvtl) [inline idl/hub_to_agent.proto:37]: nit: This message(and the corresponding rpc call and req/reply fields) should probably be called `DeleteDataDirectories`.  It's not part of this PR, though.

File: upgrade/directories.go (hunk 6 above)
Comment (bhuvnesh2703) [inline upgrade/directories.go:67]: now its not only postgresFiles, it could either be the state files? so we should use a generic name.

Comment (dkrieger-pvtl) [reply upgrade/directories.go:67]: Also, this function is generic, so a name like `missingFilesError` would likely be better...the state dir isn't a postgres dir...

File: agent/delete_directories.go [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+var stateDirectoryFiles = []string {"config.json", "status.json"}
+
+func (s *Server) DeleteStateDirectory(ctx context.Context, in *idl.DeleteStateDirectoryRequest) (*idl.DeleteStateDirectoryReply, error) {
+	gplog.Info("got a request to delete the state directory from the hub")
+
+	mErr := upgrade.DeleteDirectories([]string{in.Directory}, stateDirectoryFiles)
+	return &idl.DeleteStateDirectoryReply{}, mErr.ErrorOrNil()
+}
+
+func (s *Server) DeleteDirectories(ctx context.Context, in *idl.DeleteDirectoriesRequest) (*idl.DeleteDirectoriesReply, error) {
```
Comment (bhuvnesh2703) [inline agent/delete_directories.go:22]: do you want the name to be `DeleteDataDirectories` similar to DeleteStateDirectory so that its obvious from the name what it does..?

File: agent/delete_directories_test.go (hunk 5 above)
Comment (kalensk) [reply agent/delete_directories_test.go:15]: I think I would prefer this as well. Seems quite simple and "low cost" to have a temp dir, and more "realistic".

File: upgrade/directories_test.go [hunk 12]
```diff
@@ -48,3 +55,112 @@ func ExampleTempDataDir() {
 	// /data/standby.AAAAAAAAAAA
 	// /data/primary/seg.AAAAAAAAAAA.3
 }
+
+func TestDeleteDataDirectories(t *testing.T) {
```
Comment (dkrieger-pvtl) [inline upgrade/directories_test.go:59]: Consider creating actual directories here to test...you can use a temp dir and create the files you need underneath.  A strict mock-style test pattern would mock out the system calls, but in this case, you can test the actual functionality directly in a straightforward manner.

File: cli/commands/commands.go [hunk 13]
```diff
@@ -109,6 +111,7 @@ func BuildRootCommand() *cobra.Command {
 	root.AddCommand(initialize())
 	root.AddCommand(execute())
 	root.AddCommand(finalize())
+	root.AddCommand(revert())
```
Comment (bhuvnesh2703) [inline cli/commands/commands.go:114]: how do we know what we plan to revert, the work done by initialize or the work done by execute?

File: cli/commanders/steps.go [hunk 14]
```diff
@@ -157,6 +156,28 @@ func Finalize(client idl.CliToHubClient, verbose bool) error {
...
+	fmt.Println("Revert in progress.")
+	fmt.Println()
+
+	stream, err := client.Revert(context.Background(), &idl.RevertRequest{})
+	if err != nil {
+		gplog.Error(err.Error())
+		return err
+	}
+
+	_, err = UILoop(stream, false)
```
Comment (dkrieger-pvtl) [inline cli/commanders/steps.go:170]: maybe add a // TODO: add verbose option

File: cli/commanders/steps.go [hunk 15]
```diff
@@ -157,6 +156,28 @@ func Finalize(client idl.CliToHubClient, verbose bool) error {
...
+	if err != nil {
+		return xerrors.Errorf("Revert: %w", err)
+	}
+
+	fmt.Println("")
+	fmt.Printf("The target cluster is now restored to its original state. [more information TBD]\n")
```
Comment (dkrieger-pvtl) [inline cli/commanders/steps.go:176]: There is no longer a concept of a target cluster after a successful revert.  Do you mean to say _The source cluster_...?

File: test/revert.bats [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    while read -r datadir; do
+        if [ -d $(echo "${datadir}" | tr -d '"') ]; then
+            echo "expected datadir ${datadir} to have been deleted"
+            exit 1
+        fi
+    done <<< "${target_datadirs}"
+}
+
+@test "the state directory gets deleted" {
```
Comment (bhuvnesh2703) [inline test/revert.bats:67]: instead of doing the initialize 3 times (with stop-before-cluster-creation/or not) why not do it once and have all the validations...

File: hub/delete_data_directories.go [hunk 17]
```diff
@@ -12,20 +12,36 @@ import (
 )
 
 func DeleteMirrorAndStandbyDirectories(agentConns []*Connection, cluster *greenplum.Cluster) error {
+	return DeleteDataDirectories(agentConns, cluster, false)
+}
+
+func DeleteSegmentAndStandbyDirectories(agentConns []*Connection, cluster *greenplum.Cluster) error {
```
Comment (dkrieger-pvtl) [inline hub/delete_data_directories.go:18]: We don't have a great vocabulary for describing the role of segments, but the intent here is to delete ALL data directories other than master.  So I think something like `DeleteAllDataDirectoriesButNotMaster` is clearer.  The current name leaves me wondering what is left after it finishes.

File: ci/tasks/install-tests.yml [hunk 18]
```diff
@@ -4,7 +4,7 @@ image_resource:
   type: docker-image
   source:
     repository: pivotaldata/centos-gpdb-dev
-    tag: "6-gcc6.2-llvm3.7"
+    tag: "7-gcc6.2-llvm3.7"
```
Comment (dkrieger-pvtl) [inline ci/tasks/install-tests.yml:7]: Is this meant to be a diff as part of this PR?

File: cli/commanders/steps.go (hunk 15 above)
Comment (jchampio) [inline cli/commanders/steps.go:176]: There's a sample output linked to in the story, or is that for a follow-up PR?

File: cli/commands/commands.go (hunk 13 above)
Comment (dkrieger-pvtl) [reply cli/commands/commands.go:114]: `gpupgrade revert` should undo all of the work involved in the current upgrade, whether part of initialize or execute.  The result should be the original source cluster.  For finalize, I think we require manual user intervention to recover the source, as it's "gone".

File: cli/commands/commands.go [hunk 19]
```diff
@@ -447,6 +450,40 @@ func finalize() *cobra.Command {
...
+		Long:  "reverts the upgrade and returns the cluster to its original state",
+		Run: func(cmd *cobra.Command, args []string) {
+			client := connectToHub()
+			err := commanders.Revert(client)
+			if err != nil {
+				gplog.Error(err.Error())
+				os.Exit(1)
+			}
+
+			// todo: is there a way to instruct the hub to interrupt its own process gracefully?
```
Comment (jchampio) [inline cli/commands/commands.go:466]: Isn't that what we added `hub.Server.StopServices()` to do?

File: hub/delete_state_directories_test.go [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/greenplum-db/gp-common-go-libs/testhelper"
+	"github.com/greenplum-db/gpupgrade/hub"
+	"github.com/greenplum-db/gpupgrade/idl"
+	"github.com/greenplum-db/gpupgrade/idl/mock_idl"
+	"github.com/greenplum-db/gpupgrade/utils"
+	"github.com/hashicorp/go-multierror"
+	"golang.org/x/xerrors"
```
Comment (jchampio) [inline hub/delete_state_directories_test.go:14]: `goimports`, here and elsewhere.

File: hub/delete_data_directories.go [hunk 21]
```diff
@@ -12,20 +12,36 @@ import (
...
-				(seg.Role == greenplum.MirrorRole)
+		filterFunc := func(seg *greenplum.SegConfig) bool {
+			if seg.Hostname != conn.Hostname {
+				return false
+			}
+
+			// we don't have mirrors at this point after initialize, but we may have them later
+			// if the finalize step crashes after adding mirrors or a standby
+			if includePrimaries {
+				return !(seg.ContentID == -1 && seg.Role == greenplum.PrimaryRole)
```
Comment (dkrieger-pvtl) [inline hub/delete_data_directories.go:37]: nit: `!seg.IsMaster()` might be clearer here.

File: hub/delete_state_directories.go [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+	for _, conn := range agentConns {
+		conn := conn
+
+
+		if conn.Hostname == masterHostName {
+			continue
+		}
+
+		wg.Add(1)
+		go func(c *Connection) {
```
Comment (jchampio) [inline hub/delete_state_directories.go:29]: `conn` is already pinned in scope.

File: hub/delete_state_directories.go [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+	"context"
+	"sync"
+
+	"github.com/greenplum-db/gp-common-go-libs/gplog"
+	"github.com/greenplum-db/gpupgrade/idl"
+	"github.com/greenplum-db/gpupgrade/utils"
+	"github.com/hashicorp/go-multierror"
+)
+
+//func (s *Server) DeleteDirectories(ctx context.Context, in *idl.DeleteDirectoriesRequest) (*idl.DeleteDirectoriesReply, error) {
```
Comment (dkrieger-pvtl) [inline hub/delete_state_directories.go:13]: nit: remove this line and the next one.

File: hub/revert.go [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+
+	err = s.StopAgents()
```
Comment (jchampio) [inline hub/revert.go:37]: Speaking of `StopServices()`, this seems like a duplication of that logic.

File: hub/delete_state_directories.go [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+//func (s *Server) DeleteDirectory(ctx context.Context, in *idl.DeleteDirectoryRequest) (*idl.DeleteDirectoryReply, error) {
+func DeleteStateDirectories(agentConns []*Connection, masterHostName string) error {
+	wg := sync.WaitGroup{}
+	errChan := make(chan error, len(agentConns))
+	stateDir := utils.GetStateDir()
+
+	for _, conn := range agentConns {
+		conn := conn
+
```
Comment (dkrieger-pvtl) [inline hub/delete_state_directories.go:22]: nit: remove this blank line.

File: hub/delete_state_directories.go [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+
+		if conn.Hostname == masterHostName {
+			continue
+		}
```
Comment (jchampio) [inline hub/delete_state_directories.go:26]: Some explanation of why would be good.

File: hub/delete_state_directories_test.go [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+			}
+
+			err := hub.DeleteStateDirectories(agentConns, "master")
+			if err != nil {
+				t.Errorf("unexpected err %#v", err)
+			}
+		})
+	})
+
+	t.Run("DeleteDataDirectories", func(t *testing.T) {
```
Comment (dkrieger-pvtl) [inline hub/delete_state_directories_test.go:54]: This should be "DeleteStateDirectories"

File: hub/revert.go (hunk 8 above)
Comment (dkrieger-pvtl) [reply hub/revert.go:35]: Yes, we are in control of that, but how do we design the code to make it easy to do?  It's a dynamic thing...any developer needs to know which function can be called "after revert".  For instance, if I were working on the internals of the next call(`s.StopAgents`) as part of the upcoming pidfile work, how would I know this?  Maybe an annotation-like comment?  There shouldn't be too many things we do after revert.

File: cli/commands/commands.go (hunk 13 above)
Comment (kalensk) [reply cli/commands/commands.go:114]: This PR is just a first step towards a fully functional customer facing revert command.

**Comment by dkrieger-pvtl** [comment]
Overall, I think this PR is well-designed and tested.  But I have a lot of code comments/suggestions(see comments I've made).  I also haven't looked at the tests yet.

File: test/revert.bats (hunk 16 above)
Comment (kalensk) [reply test/revert.bats:67]: I was thinking the same thing...

File: cli/commands/commands.go [hunk 28]
```diff
@@ -447,6 +450,40 @@ func finalize() *cobra.Command {
...
+			err = exec.Command("pkill", "-f", "gpupgrade hub").Run()
```
Comment (dkrieger-pvtl) [inline cli/commands/commands.go:467]: I wonder if the technique in agent/server.go::StopAgent() used to gracefully stop the agent will work here.  Although I think we might want to switch to pid files and a process-level lifecycle for the hub and agents anywa.

File: hub/delete_data_directories.go (hunk 17 above)
Comment (kalensk) [reply hub/delete_data_directories.go:18]: FWIW, I tend prefer to phrase things _explicitly_ and in the positive such as `DeleteSegmentAndStandbyDataDirectories`. This way it makes it explicit and clear what is being operated on, rather than "notMaster" which is more implicit and opaque to me.

File: ci/tasks/install-tests.yml (hunk 18 above)
Comment (kalensk) [reply ci/tasks/install-tests.yml:7]: I believe it was to include `jq` for testing, but that is under discussion in comment https://github.com/greenplum-db/gpupgrade/pull/310#discussion_r406885951. I am not sure if jq is needed in other places.

File: hub/delete_data_directories.go (hunk 21 above)
Comment (kalensk) [reply hub/delete_data_directories.go:37]: +1

File: hub/revert.go [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/greenplum-db/gp-common-go-libs/gplog"
+	"github.com/greenplum-db/gpupgrade/upgrade"
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/greenplum-db/gpupgrade/idl"
+	"github.com/greenplum-db/gpupgrade/step"
+)
+
+func (s *Server) Revert(_ *idl.RevertRequest, stream idl.CliToHub_RevertServer) (err error) {
+	st, err := step.Begin(s.StateDir, "revert", stream)
```
[the remark this answers is not in the document]
Comment (kalensk) [reply hub/revert.go:15]: Agreed. I am not sure trying to make this fit the` step.Steap` pattern is helping us here, and in fact may not actually be helping.

File: hub/delete_state_directories.go (hunk 25 above)
Comment (jchampio) [reply hub/delete_state_directories.go:22]: (run the autoformatter please)

**Comment by marksliva** [comment]
Closing this PR, and making some structural changes before opening a new PR.

File: cli/commands/commands.go (hunk 28 above)
Comment (marksliva) [reply cli/commands/commands.go:467]: I ended up re-using the `StopServices` that `gpupgrade kill-services` uses, which was Jacob's suggestion.

File: hub/revert.go (hunk 24 above)
Comment (marksliva) [reply hub/revert.go:37]: Fixed.

File: hub/revert.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (marksliva) [reply hub/revert.go:35]: Done.

File: hub/revert.go [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	// todo: DeleteStateDirectories()
+
+	err = s.StopAgents()
+	if err != nil {
+		return err
+	}
+
+	err = upgrade.DeleteDirectories([]string{s.Config.Target.MasterDataDir()}, nil)
```
[the remark this answers is not in the document]
Comment (marksliva) [reply hub/revert.go:42]: Not exactly. The latest implementation passes in `PostgresFiles` for added safety.

File: test/revert.bats [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+        --target-bindir="$GPHOME/bin" \
+        --source-master-port="${PGPORT}" \
+        --temp-port-range 6020-6040 \
+        --disk-free-ratio 0 \
+        --stop-before-cluster-creation \
+        --verbose 3>&-
+
+    gpupgrade revert
+
+    if [ -d $(echo "${GPUPGRADE_HOME}") ]; then
```
[the remark this answers is not in the document]
Comment (marksliva) [reply test/revert.bats:79]: Fixed.

File: test/revert.bats [hunk 32]
```diff
@@ -0,0 +1,83 @@
...
+        --disk-free-ratio 0 \
+        --verbose 3>&-
+
+    # parse config.json for the datadirs
+    local target_datadirs=$(jq ".Target.Primaries[].DataDir" "${GPUPGRADE_HOME}/config.json")
+
+    gpupgrade revert
+
+    while read -r datadir; do
+        if [ -d $(echo "${datadir}" | tr -d '"') ]; then
```
[the remark this answers is not in the document]
Comment (marksliva) [reply test/revert.bats:60]: Done.

File: test/revert.bats [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+        --temp-port-range 6020-6040 \
+        --disk-free-ratio 0 \
+        --stop-before-cluster-creation \
+        --verbose 3>&-
+
+    if [[ $(process_is_running "[g]pupgrade hub") -eq 0 ]]; then
+        # todo: this makes a single host assumption
+        echo 'expected hub to be running'
+        exit 1
+    fi
```
[the remark this answers is not in the document]
Comment (marksliva) [reply test/revert.bats:27]: Fixed.

File: agent/delete_directories.go [hunk 34]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"context"
+
+	"github.com/greenplum-db/gp-common-go-libs/gplog"
+	"github.com/greenplum-db/gpupgrade/idl"
+	"github.com/greenplum-db/gpupgrade/upgrade"
+)
+
+// todo: add back the [...]
```
[the remark this answers is not in the document]
Comment (marksliva) [reply agent/delete_directories.go:11]: Done. Now I am passing these slices in like this: `stateDirectoryFiles[:]`.

File: hub/delete_state_directories.go (hunk 25 above)
Comment (marksliva) [reply hub/delete_state_directories.go:22]: Done.

File: hub/delete_state_directories.go (hunk 23 above)
Comment (marksliva) [reply hub/delete_state_directories.go:13]: Done.

File: hub/delete_data_directories.go (hunk 21 above)
Comment (marksliva) [reply hub/delete_data_directories.go:37]: Good catch. Done.

File: ci/tasks/install-tests.yml (hunk 18 above)
Comment (marksliva) [reply ci/tasks/install-tests.yml:7]: Yes, this was done so we could use `jq`, which is isolated to just `revert.bats` for now.

File: hub/delete_data_directories.go (hunk 17 above)
Comment (marksliva) [reply hub/delete_data_directories.go:18]: It seems fine to list the 2 things that do want to delete. If there were 3 things, saying what we are not deleting would start being a better option.

File: test/revert.bats (hunk 16 above)
Comment (marksliva) [reply test/revert.bats:67]: Done. This was the result of TDDing the feature, but since these are long running integration tests, it does make sense to consolidate them when possible.

File: cli/commanders/steps.go (hunk 15 above)
Comment (marksliva) [reply cli/commanders/steps.go:176]: Good catch.

File: cli/commanders/steps.go (hunk 14 above)
Comment (marksliva) [reply cli/commanders/steps.go:170]: Done.

File: agent/delete_directories.go (hunk 11 above)
Comment (marksliva) [reply agent/delete_directories.go:22]: Fixed.

File: upgrade/directories.go (hunk 6 above)
Comment (marksliva) [reply upgrade/directories.go:67]: Fixed.

File: idl/hub_to_agent.proto (hunk 10 above)
Comment (marksliva) [reply idl/hub_to_agent.proto:37]: This is a good idea though. It'll make things easier to follow. I changed it.

File: agent/delete_directories.go (hunk 9 above)
Comment (marksliva) [reply agent/delete_directories.go:26]: Renamed the server function. But leaving the `upgrade.DeleteDirectories` as is since it can delete either state directories or data directories.

File: hub/revert.go (hunk 8 above)
Comment (marksliva) [reply hub/revert.go:35]: I had commented this mainly because I was still thinking about how it would be tested in a multi-host setting. Which it turns out we can use the multi-host vagrant code to do that. I added it back in.

File: upgrade/directories.go (hunk 6 above)
Comment (marksliva) [reply upgrade/directories.go:67]: The reason it is a separate multi error from `mErr` is so that it can catch all missing files, and if it finds at least one missing file, all the errors get appended to `mErr`. Either way, I renamed this to `missingFilesError`.

File: idl/hub_to_agent.proto (hunk 4 above)
Comment (marksliva) [reply idl/hub_to_agent.proto]: Done.

File: agent/delete_data_directories.go (hunk 3 above)
Comment (marksliva) [reply agent/delete_data_directories.go]: Done.

File: test/revert.bats (hunk 1 above)
Comment (marksliva) [reply test/revert.bats:56]: The jq command can be written like this to extract hostname info as well:
`jq -r '.Target.Primaries[] | .Hostname + " " + .DataDir' "${GPUPGRADE_HOME}/config.json"`
So that can still work.

File: hub/delete_master_data_directory.go (hunk 2 above)
Comment (marksliva) [reply hub/delete_master_data_directory.go]: This has been deduplicatd.

File: cli/commands/commands.go (hunk 19 above)
Comment (marksliva) [reply cli/commands/commands.go:466]: Good point. Initially, we thought we needed to control the stopping of the services before removing the state dirs. But it turns out we can reuse the `StopServices` logic.

File: hub/delete_state_directories.go (hunk 26 above)
Comment (marksliva) [reply hub/delete_state_directories.go:26]: Done.

File: hub/delete_state_directories_test.go (hunk 20 above)
Comment (marksliva) [reply hub/delete_state_directories_test.go:14]: Fixed.

Comment (marksliva) [reply hub/delete_state_directories_test.go:14]: Fixed.

File: agent/delete_directories_test.go (hunk 5 above)
Comment (marksliva) [reply agent/delete_directories_test.go:15]: Done.

> > The revert command is only being tested after running a full initialize,
> > however it should be idempotent enough to handle reverting from any substep or
> > failure encountered during initialize.
> 
> I think you mean to say that this story is only handling revert after a full initialize has completed, right? Recovering from a failure mid-initialize is more work(might have to stop postgres processes, etc), and will be handled later. As will saving off the state directory.
> 
> Note the gpAdminLogs are not touched by this PR, but I suppose we consider those system logs and hence we do not need to delete them.

Sure. It will be useful to do a more thorough pass at idempotence at some later point. I updated the commit message to reflect this.

File: upgrade/directories_test.go (hunk 7 above)
Comment (marksliva) [reply upgrade/directories_test.go:139]: Since this test is now integrated to the file system, this comment is no longer relevant.

File: cli/commands/commands.go (hunk 13 above)
Comment (marksliva) [reply cli/commands/commands.go:114]: Yes. In the long term, `revert` will be able to revert from any point up until finalize. This PR is just handling revert after initialize.
